App/src/Service/GroupExpenseBalancer.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace App\Service;
+
+use App\Entity\Bilan;
+use App\Entity\Expense;
+
+class GroupExpenseBalancer
+{
+    public function expenseBalancer($expenses)
+    {
+        /**
+         * @var Bilan[]
+         */
+        $bilans = array_reduce($expenses, static function (array $bilans, Expense $expense) {
+            $payer = $expense->getPayer();
+            if (!array_key_exists($payer, $bilans)) {
+                return [...$bilans, $payer => new Bilan($payer)];
+            }
+
+            return $bilans;
+        }, []);
+
+        $this->setExpenses($expenses, $bilans);
+
+        return $bilans;
+    }
+
+    private function setExpenses($expenses, $bilans)
+    {
+        foreach ($expenses as $expense) {
+            $amount = $expense->getAmount();
+            $participants = $expense->getParticipants();
+            $countParticipants = count($participants);
+            $rest = $amount % $countParticipants;
+            $amountByParticipants = ($amount - $rest) / $countParticipants;
+            $payer = $expense->getPayer();
+
+            $this->updateBilan($bilans, $amount, $participants, $payer, $amountByParticipants);
+        }
+    }
+
+    private function updateBilan($bilans, $amount, $participants, $payer, $amountByParticipants)
+    {
+        foreach ($bilans as $bilan) {
+            $name = $bilan->getName();
+            $cost = $bilan->getCost();
+            $participations = $bilan->getParticipation();
+            $owe = $bilan->getOwe();
+
+            if ($payer === $name) {
+                $bilan->setCost($cost + $amount);
+            }
+            if (in_array($name, $participants)) {
+                $bilan->setParticipation($participations + $amountByParticipants);
+            }
+
+            foreach ($participants as $participant) {
+                if ($name === $payer && $name !== $participant) {
+                    if (array_key_exists($participant, $owe)) {
+                        $owe[$participant] += $amountByParticipants;
+                    }
+                    $owe[$participant] = $amountByParticipants;
+                }
+            }
```

Comment:
When you have `if` nesting, you can use the [early return pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8) to improve the readability 😉

Reply:
```
foreach ($participants as $participant) {
                if ($name !== $payer) {
                    break;
                }
                if ($name !== $participant) {
                    $owe[$participant] = ($owe[$participant] ?? 0) + $amountByParticipants;
                }
```

Something like this ?

Reply:
Ta suggestion marchera mais ce n'est pas le strict équivalent de ton premier jet. Pour avoir quelque chose d'équivalent il faut prendre tes conditions une par une et faire une négation. Dans ton cas cela te donnerait quelque chose comme ceci :

```php
            foreach ($participants as $participant) {
                if ($name !== $payer || $name === $participant) {
                    continue;
                }

                if (array_key_exists($participant, $owe)) {
                    $owe[$participant] += $amountByParticipants;
                }

                $owe[$participant] = $amountByParticipants;
            }
```

Reply:
Okay, done !